File: cpp-alg/leecode/Q283_move_zeroes.cpp
#include <iostream>
#include <vector>

void moveZeroes(std::vector<int> &nums) {
  // for (int i = 0; i < nums.size(); i++) {
  //  for (int j = i + 1; j < nums.size(); j++) {
  //  if (nums[i] == 0) {
  //  int temp = nums[i];
  //  nums[i] = nums[j];
  //  nums[j] = temp;
  //     }
  //   }
  // }

  // track non-zero numbers
  int count = 0;
  for (int i = 0; i < nums.size(); i++) {
    if (nums[i] != 0)
      nums[count++] = nums[i];
  }
  // replace other slots to 0
  for (int i = count; i < nums.size(); i++) {
    nums[i] = 0;
  }
}

int main() {
  std::vector<int> nums = {1, 0, 3, 4, 0, 9, 10};
  moveZeroes(nums);
  for (int num : nums)
    std::cout << num << " ";
}

File: python-alg/leecode/q283_move_zeroes.py
def move_zeroes(nums):
    index = 0
    count = 0
    for i in range(len(nums)):
        if nums[i] != 0:
            nums[index] = nums[i]
            index += 1
        else:
            count += 1
    start_index = max(0, len(nums) - count)
    for i in range(start_index, len(nums)):
        nums[i] = 0

    return nums


nums = [1, 0, 1]
move_zeroes(nums)
print(nums)
